### ISSUE TITLE
 Flux gradient returns "nothing" while using SliceMap.mapcols in loss function   

### ISSUE THREAD

**Issue by 5tinzi** [mcabbott/SliceMap.jl#16]
Hi,  I found that with `SliceMap.mapcols` Zygote/Flux gradients will return nothing.  I'm not sure if this is the correct usage of `SliceMap.mapcols`.   Could you give some suggest? 

[what I expect ]:  get the proper gradients  while using `mapcols`  with Zygote/Flux.

A demo example to repeat:
```julia
using Flux, SliceMap
model = Chain(
    Dense(5,32),
    Dense(32,5)
)

X = rand(5,5)
Y = 1:5
# a loss function with mapcols
function lossA(X)
    SliceMap.mapcols(x -> Flux.Losses.mse(model(x),Y), X) |> mean
end
# another loss function without mapcols
function lossB(X)
    loss_ = 0
    for i in 1:size(X,2)
        loss_ = loss_  + Flux.Losses.mse(model(X[:,i]),Y)
    end 
    loss_ / size(X,2)
end

# train the model  with  loss function A
grads = Flux.gradient(Flux.params(model)) do
    lossA(X)
end;
grads.grads # all grad is nothing 
# IdDict{Any, Any} with 5 entries:
#   Float32[0.0, 0.0, 0.0, 0.0, 0.0]           => nothing
#   Float32[0.327589 -0.112092 … 0.327733 -0.… => nothing
#   Float32[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0… => nothing
#   Float32[0.178549 -0.265913 … -0.274821 -0… => nothing


# train the model  with  loss function B. 
grads = Flux.gradient(Flux.params(model)) do
    lossB(X)
end;
grads.grads # it works.
# IdDict{Any, Any} with 6 entries:
#   Float32[0.0, 0.0, 0.0, 0.0, 0.0]           => Float32[-0.503307, -0.799109, -1.07115, -1.68975, -2.00948]
#   Float32[0.327589 -0.112092 … 0.327733 -0.… => Float32[0.0582891 0.0484881 … -0.145161 0.0146028; 0.0890983 0.0792296 … -0.228783 …
#   Float32[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0… => Float32[0.241087, -0.283567, -0.942893, -0.385205, -0.549046, -0.852832, 0.125945, …
#   Float32[0.178549 -0.265913 … -0.274821 -0… => Float32[0.124569 0.0878892 … 0.0644054 

``` 
Thank you for any suggestion.

**Comment by mcabbott** [comment]
This is unfortunately expected. It's what this warning is saying:

https://github.com/mcabbott/SliceMap.jl/blob/master/src/SliceMap.jl#L29

The function `f` is here `x -> Flux.Losses.mse(model(x),Y)` which contains all the parameters, but `mapcols` isn't smart enough to understand that.

This could be fixed I think. But probably more productive now is just to write a rule to make Zygote etc. understand the built-in `mapslices` etc.
